Based on:
https://nanoporetech.github.io/modkit/intro_dmr.html#perform-differential-methylation-scoring

Select kernal: dimelo_v2_modkit_parsing

Preparing the input data
The inputs to all modkit dmr commands are two or more bedMethyl files (created by modkit pileup) that have been compressed with bgzip and indexed with tabix. An example of how to generate the input data is shown below:


ref=grch38.fasta
threads=32

norm=normal_sample.bam
norm_pileup=normal_pileup.bed

modkit pileup ${norm} ${norm_pileup} \
  --cpg \
  --ref ${ref} \
  --threads ${threads} \
  --log-filepath log.txt

bgzip -k ${norm_pileup}
tabix -p bed ${norm_pileup}.gz

# pileup and compression can also be done in one step
tumor=tumor_sample.bam
tumor_pileup=tumor_pileup.bed.gz

modkit pileup ${tumor} - \
  --cpg \
  --ref ${ref} \
  --threads ${threads} \
  --log-filepath log.txt | ${bgzip} -c > ${tumor_pileup}

tabix -p bed ${tumor_pileup}

In [1]:
%%bash
echo "hello"

hello


In [2]:
# %%bash

# ref="/home/michalula/data/ref_genomes/to_t2t_v1_1/chm13.draft_v1.1.fasta"
# threads=32

# CROFF_day28_bam="/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_CROFF_Day28/mergered_outputs/sort_merge_sort_align_t2t_v1_1_trim_20250721_nCATs_Tcells_CROFF_Day28_minion_run2_day8.dna_r9.4.1_e8_sup@v3.3.5mCG.bam"
# pileup_CROFF_day28_bam="/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_CROFF_Day28/mergered_outputs/pileup_sort_merge_sort_align_t2t_v1_1_trim_20250721_nCATs_Tcells_CROFF_Day28_minion_run2_day8.dna_r9.4.1_e8_sup@v3.3.5mCG.bam"

# # norm=normal_sample.bam
# # norm_pileup=normal_pileup.bed

# modkit pileup ${CROFF_day28_bam} ${pileup_CROFF_day28_bam} \
#   --cpg \
#   --ref ${ref} \
#   --threads ${threads} \
#   --log-filepath log.txt

# bgzip -k ${pileup_CROFF_day28_bam}
# tabix -p bed ${pileup_CROFF_day28_bam}.gz

In [3]:
! ls /home/michalula/data/ref_genomes/t2t_v2_0/

chm13v2.0.fa	  chm13v2.0.fa.fai		   haplotype_vcf
chm13v2.0.fa.amb  chm13v2.0.fa.pac		   up_chm13v2.0.fasta
chm13v2.0.fa.ann  chm13v2.0.fa.sa		   up_chm13v2.0.fasta.fai
chm13v2.0.fa.bwt  convert_to_uppercase_fasta.bash


In [4]:
! ls /home/michalula/data/cas9_nanopore/data/20250908_nCATs_T_CRoff_Day_35/5mCG/to_t2t_v2_0

align_t2t_v2_0_trim_20250908_Day35_CROFF_Tcells_2Libraries_Minion_R9.dna_r9.4.1_e8_sup@v3.3.5mCG.bam
sort_align_t2t_v2_0_trim_20250908_Day35_CROFF_Tcells_2Libraries_Minion_R9.dna_r9.4.1_e8_sup@v3.3.5mCG.bam
sort_align_t2t_v2_0_trim_20250908_Day35_CROFF_Tcells_2Libraries_Minion_R9.dna_r9.4.1_e8_sup@v3.3.5mCG.bam.bai
summary_sort_align_t2t_v2_0_trim_20250908_Day35_CROFF_Tcells_2Libraries_Minion_R9.dna_r9.4.1_e8_sup@v3.3.5mCG.tsv


In [6]:
%%bash

ref_genome_fa="/home/michalula/data/ref_genomes/t2t_v2_0/chm13v2.0.fa"
# ref= "/home/michalula/data/ref_genomes/to_t2t_v1_1/chm13.draft_v1.1.fasta"

threads=32

date_today="20251109"
data_folder_path="/home/michalula/data/cas9_nanopore/data/20250908_nCATs_T_CRoff_Day_35/5mCG/to_t2t_v2_0/"
CROFF_day35_bam=${data_folder_path}"sort_align_t2t_v2_0_trim_20250908_Day35_CROFF_Tcells_2Libraries_Minion_R9.dna_r9.4.1_e8_sup@v3.3.5mCG.bam"
pileup_CROFF_day35_bam=${data_folder_path}{date_today}"_pileup_CROFF_Day35_Tcells.bam"
# "/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_CROFF_Day28/mergered_outputs/pileup_sort_merge_sort_align_t2t_v1_1_trim_20250721_nCATs_Tcells_CROFF_Day28_minion_run2_day8.dna_r9.4.1_e8_sup@v3.3.5mCG.bam"
# norm=normal_sample.bam
# norm_pileup=normal_pileup.bed

modkit pileup ${CROFF_day35_bam} ${pileup_CROFF_day35_bam} \
  --cpg \
  --ref ${ref_genome_fa} \
  --threads ${threads} \
  --log-filepath log.txt

bgzip -k ${pileup_CROFF_day35_bam}
tabix -p bed ${pileup_CROFF_day35_bam}.gz


> calculated chunk size: 48, interval size 100000, processing 4800000 positions concurrently
> filtering to only CpG motifs
> attempting to sample 10042 reads
> Using filter threshold 0.7597656 for C.
> Done, processed 11762972 rows. Processed ~129977 reads and skipped ~150 reads.


<!>
> Threshold of  0.7597656 for base C is low. Consider increasing the filter-percentile or specifying a higher threshold.
> Done, processed 11762972 rows. Processed ~129977 reads and skipped ~150 reads.

In [ ]:
%%bash

mkdir ${data_folder_path}"pileup"
# pileup_dir_path="/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/5mCG/to_t2t_v2_0/pileup"
# mkdir ${pileup_dir_path}

Unedit_day28_bam="/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/5mCG/to_t2t_v2_0/sort_align_t2t_v2_0_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged_flowcells_output.dna_r9.4.1_e8_sup@v3.3.5mCG.bam"
# pileup_Unedit_day28_bam="/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup@v3.3.5mCG.bam"
# pileup_Unedit_day28_bam=${pileup_dir_path}"/pileup_Unedit_Day28_Tcells_20250721.bam"
pileup_Unedit_day28_bam="/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/5mCG/to_t2t_v2_0/pileup/pileup_Unedit_Day28_Tcells_20250721.bam"

# "/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup.5mCG.bam"
  
ref_genome_fa="/home/michalula/data/ref_genomes/t2t_v2_0/chm13v2.0.fa"
# ref="/home/michalula/data/ref_genomes/to_t2t_v1_1/chm13.draft_v1.1.fasta"

threads=32
 

modkit pileup ${Unedit_day28_bam} ${pileup_Unedit_day28_bam} \
  --cpg \
  --ref ${ref_genome_fa} \
  --threads ${threads} \
  --log-filepath log.txt

bgzip -k ${pileup_Unedit_day28_bam}
tabix -p bed ${pileup_Unedit_day28_bam}.gz

# modkit pileup ${Unedit_day28_bam} ${pileup_Unedit_day28_bam} \
#   --cpg \
#   --ref ${ref} \
#   --threads ${threads} \
#   --log-filepath log.txt

# bgzip -k ${pileup_Unedit_day28_bam}
# tabix -p bed ${pileup_Unedit_day28_bam}.gz

mkdir: cannot create directory ‘/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/5mCG/to_t2t_v2_0/pileup’: File exists
> calculated chunk size: 48, interval size 100000, processing 4800000 positions concurrently
> filtering to only CpG motifs
> attempting to sample 10042 reads
> failed to get modbase info for record 02866349-9f01-4302-9a1c-708042ab1756, Bad Input: invalid MM delta list, Parsing Error: Error { input: "", code: Digit }
> failed to get modbase info for record 81342b67-3ffb-432f-967c-cbd4f5f6fd23, Bad Input: invalid MM delta list, Parsing Error: Error { input: "", code: Digit }
> failed to get modbase info for record c6a4eb69-74f6-4cc7-a0b2-a818bef21c3a, Bad Input: invalid MM delta list, Parsing Error: Error { input: "", code: Digit }
> failed to get modbase info for record d64f46b8-658e-44ca-a77d-f3ad558b0a1e, Bad Input: invalid MM delta list, Parsing Error: Error { input: "", code: Digit }
> failed to get modbase info for record

In [ ]:
# %%bash

# Unedit_day28_bam="/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup@v3.3.5mCG.bam"
# # pileup_Unedit_day28_bam="/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup@v3.3.5mCG.bam"
# pileup_Unedit_day28_bam="/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup.5mCG.bam"
  
# ref="/home/michalula/data/ref_genomes/to_t2t_v1_1/chm13.draft_v1.1.fasta"
# threads=32
 
# modkit pileup ${Unedit_day28_bam} ${pileup_Unedit_day28_bam} \
#   --cpg \
#   --ref ${ref} \
#   --threads ${threads} \
#   --log-filepath log.txt

# bgzip -k ${pileup_Unedit_day28_bam}
# tabix -p bed ${pileup_Unedit_day28_bam}.gz

In [ ]:
# %%bash

# Unedit_day28_bam="/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup@v3.3.5mCG.bam"
# # pileup_Unedit_day28_bam="/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup@v3.3.5mCG.bam"
# pileup_Unedit_day28_bam="/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup.5mCG.bam"
  
# ref="/home/michalula/data/ref_genomes/to_t2t_v1_1/chm13.draft_v1.1.fasta"
# threads=32
 
# modkit pileup ${Unedit_day28_bam} ${pileup_Unedit_day28_bam} \
#   --cpg \
#   --ref ${ref} \
#   --threads ${threads} \
#   --log-filepath log.txt

# bgzip -k ${pileup_Unedit_day28_bam}
# tabix -p bed ${pileup_Unedit_day28_bam}.gz

In [ ]:
# %%bash
# # pileup and compression can also be done in one step

# ref="/home/michalula/data/ref_genomes/to_t2t_v1_1/chm13.draft_v1.1.fasta"
# threads=32

# Unedit_day28_bam="/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup@v3.3.5mCG.bam"
# pileup_Unedit_day28_bam_gz="/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup@v3.3.5mCG.bam.gz"
 
# # tumor=tumor_sample.bam
# # tumor_pileup=tumor_pileup.bed.gz

# # modkit pileup ${Unedit_day28_bam} - \
# #   --cpg \
# #   --ref ${ref} \
# #   --threads ${threads} \
# #   --log-filepath log.txt | ${bgzip} -c > ${pileup_Unedit_day28_bam}

# # tabix -p bed ${pileup_Unedit_day28_bam}

3. Detecting differential modification at single base positions
The modkit dmr pair command has the ability to score individual bases (e.g. differentially methylated CpGs). To run single-base analysis on one or more paired samples, simply omit the --regions (-r) option when running modkit dmr pair. When performing single-base analysis the likelihood ratio score and a MAP-based p-value are available. For details on the likelihood ratio score and the MAP-based p-value, see the scoring details section. For example the above command becomes:

dmr_result=single_base_haplotype_dmr.bed

modkit dmr pair \
  -a ${hp1_pileup}.gz \
  -b ${hp2_pileup}.gz \
  -o ${dmr_result} \
  --ref ${ref} \
  --base C \
  --threads ${threads} \
  --log-filepath dmr.log

In [ ]:
! ls "/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/5mCG"

20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged_flowcells_output.dna_r9.4.1_e8_sup@v3.3.5mCG.bam
t2t_v1_1
to_t2t_v2_0
trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged_flowcells_output.dna_r9.4.1_e8_sup@v3.3.5mCG.bam
trim_5mCG_5hmCG_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged_flowcells_output.dna_r9.4.1_e8_sup@v3.3.5mCG.bam


In [ ]:
! ls /home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/5mCG/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup@v3.3.5mCG.bam.gz

ls: cannot access '/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/5mCG/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup@v3.3.5mCG.bam.gz': No such file or directory


In [ ]:
/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup.5mCG.bam.gz

In [ ]:
    %%bash

   ls -ld /home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs
   ls -l /home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup.5mCG.bam.gz

drwxrwxr-x 2 michalula michalula 4096 Jul 31 00:43 /home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs
-rw-rw-r-- 1 michalula michalula 5758746 Jul 31 00:43 /home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup.5mCG.bam.gz


In [ ]:
%%bash

   # chmod u+rw /home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup.5mCG.bam.gz
   # chmod u+rwx /home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs

In [ ]:
%%bash

   chmod u+rwx "/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/5mCG/to_t2t_v2_0/pileup/"
   chmod u+rwx "/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_CROFF_Day28/mergered_outputs/5mCG/to_t2t_v2_0/pileup_CROFF_Day28_Tcells_20250721.bam.gz"

In [ ]:
%%bash

ls -ld /home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs
ls -ld "/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_CROFF_Day28/mergered_outputs/5mCG/to_t2t_v2_0/pileup_CROFF_Day28_Tcells_20250721.bam.gz"

drwxrwxr-x 4 michalula michalula 4096 Aug  1 01:35 /home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs
-rwxrw-r-- 1 michalula michalula 5354303 Sep 24 12:10 /home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_CROFF_Day28/mergered_outputs/5mCG/to_t2t_v2_0/pileup_CROFF_Day28_Tcells_20250721.bam.gz


In [ ]:
# ls /home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup.5mCG.bam.gz

In [ ]:
 %%bash
 
 # 3. Detecting differential modification at single base positions
# The modkit dmr pair command has the ability to score individual bases (e.g. differentially methylated CpGs). To run single-base analysis on one or more paired samples, simply omit the --regions (-r) option when running modkit dmr pair. When performing single-base analysis the likelihood ratio score and a MAP-based p-value are available. For details on the likelihood ratio score and the MAP-based p-value, see the scoring details section. For example the above command becomes:

dmr_result="/home/michalula/code/epiCausality/epiCode/differential_methyl/modkit_output/single_base_day28_CRoff_vs_Unedit_dmr_20251109.bed"

# pileup_Unedit_day28_bam="/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup@v3.3.5mCG.bam"
# removed the @ from the file name
# pileup_Unedit_day28_bam="/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup.5mCG.bam"
# pileup_Unedit_day28_bam="/home/michalula/code/epiCausality/epiCode/differential_methyl/pileups/t2t_v1_1/pileup_Unedit_Day28_Tcells_20250721.bam"
pileup_Unedit_day28_bam="/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/5mCG/to_t2t_v2_0/pileup/pileup_Unedit_Day28_Tcells_20250721.bam"

echo "${pileup_Unedit_day28_bam}.gz"
ls -l "${pileup_Unedit_day28_bam}.gz"
# "/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup.5mCG.bam"
# pileup_Unedit_day28_bam_gz="/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/pileup_sorted_align_t2t_v1_1_trim_20250721_nCATs_Tcells_UNEDITED_Day28_minion_merged.dna_r9.4.1_e8_sup.5mCG.bam.gz"
  
# pileup_CROFF_day28_bam="/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_CROFF_Day28/mergered_outputs/pileup_sort_merge_sort_align_t2t_v1_1_trim_20250721_nCATs_Tcells_CROFF_Day28_minion_run2_day8.dna_r9.4.1_e8_sup@v3.3.5mCG.bam"
# pileup_CROFF_day28_bam="/home/michalula/code/epiCausality/epiCode/differential_methyl/pileups/t2t_v1_1/pileup_CROFF_Day28_Tcells_20250721.bam"
pileup_CROFF_day28_bam="/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_CROFF_Day28/mergered_outputs/5mCG/to_t2t_v2_0/pileup_CROFF_Day28_Tcells_20250721.bam"

ref_genome_fa="/home/michalula/data/ref_genomes/t2t_v2_0/chm13v2.0.fa"
# ref="/home/michalula/data/ref_genomes/to_t2t_v1_1/chm13.draft_v1.1.fasta"
threads=32
 
cd '/home/michalula/code/epiCausality/epiCode/differential_methyl'

modkit dmr pair \
  -a ${pileup_Unedit_day28_bam}.gz \
  -b ${pileup_CROFF_day28_bam}.gz \
  -o ${dmr_result} \
  --ref ${ref_genome_fa} \
  --base C \
  --threads ${threads} \
  --log-filepath dmr.log


/home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/5mCG/to_t2t_v2_0/pileup/pileup_Unedit_Day28_Tcells_20250721.bam.gz
-rw-rw-r-- 1 michalula michalula 7514010 Sep 24 12:17 /home/michalula/data/cas9_nanopore/data/20250721_nCATs_Tcells_UNEDITED_Day28/merged_outputs/5mCG/to_t2t_v2_0/pileup/pileup_Unedit_Day28_Tcells_20250721.bam.gz


> loading sites from input 'a' bedMethyl
> loaded 867290 sites
> loading 48 sites at a time
> 111264 regions processed successfully and 756026 regions failed


In [ ]:
%%bash

dmr_result="/home/michalula/code/epiCausality/epiCode/differential_methyl/modkit_output/single_base_day28_CRoff_vs_Unedit_dmr_20251109.bed"

ls -lah $dmr_result
# cat $dmr_result

-rw-rw-r-- 1 michalula michalula 7.8M Nov  9 12:34 /home/michalula/code/epiCausality/epiCode/differential_methyl/modkit_output/single_base_day28_CRoff_vs_Unedit_dmr_20251109.bed


In [ ]:
pwd

'/home/michalula/code/epiCausality/epiCode/differential_methyl'

In [ ]:
# Visualize modkit DMR result
import os
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.graph_objects as go

# Path to the DMR BED output (adjust if you use a different file)
dmr_path = "/home/michalula/code/epiCausality/epiCode/differential_methyl/modkit_output/single_base_day28_CRoff_vs_Unedit_dmr_20251109.bed"
out_dir = os.path.dirname(dmr_path) or "."
os.makedirs(out_dir, exist_ok=True)

# Try reading with a header first, fall back to headerless read
try:
    df = pd.read_csv(dmr_path, sep='	', comment='#', header=0, engine='python')
    if df.shape[1] < 3:
        df = pd.read_csv(dmr_path, sep='	', comment='#', header=None, engine='python')
except Exception as e:
    print("Failed to read DMR file:", e)
    raise

# If the file had no header, pandas will give integer column names - assign common BED/DMR names
if isinstance(df.columns[0], int):
    ncols = df.shape[1]
    default_names = ["chrom", "start", "end", "name", "score", "strand"]
    names = default_names[:ncols] + [f"col{i}" for i in range(len(default_names), ncols)]
    df.columns = names

# Coerce numeric-looking columns where possible
for c in df.columns:
    try:
        df[c] = pd.to_numeric(df[c], errors='ignore')
    except Exception:
        pass

print("DMR file:", dmr_path)
print("Rows,cols:", df.shape)
display(df.head())
display(df.describe(include="all").T)

# Make histograms of numeric columns and save as HTML
numcols = df.select_dtypes(include=[np.number]).columns.tolist()
if len(numcols):
    for col in numcols:
        fig = px.histogram(df, x=col, nbins=80, title=f"Histogram of {col}")
        html = os.path.join(out_dir, f"dmr_hist_{col}.html")
        fig.write_html(html)
        print("Saved", html)

# Scatter: genomic position vs score (if present)
score_col = None
if 'start' in df.columns:
    if 'score' in df.columns:
        score_col = 'score'
    else:
        # try to find a numeric column to plot against position
        for c in numcols:
            if c != 'start' and c != 'end':
                score_col = c
                break
    if score_col is not None:
        df_plot = df.copy()
        df_plot['pos'] = df_plot['start'].astype(int)
        fig = px.scatter(df_plot, x='pos', y=score_col, color='chrom' if 'chrom' in df_plot.columns else None,
                         title=f"Position vs {score_col}", hover_data=df_plot.columns.tolist(), height=600)
        html = os.path.join(out_dir, 'dmr_position_score_scatter.html')
        fig.write_html(html)
        print("Saved", html)

# Show a table of top hits (by p-value or by score if p-value not present)
rank_col = None
for key in ['pvalue','p_val','p.value','p','padj','qval','score']:
    if key in df.columns:
        rank_col = key
        break
if rank_col is None and len(numcols):
    rank_col = numcols[0]
if rank_col is not None:
    ascending = True if 'p' in str(rank_col).lower() or 'q' in str(rank_col).lower() else False
    top = df.sort_values(by=rank_col, ascending=ascending).head(50)
    display(top)

print("Done. Visualizations saved to:", out_dir)

DMR file: /home/michalula/code/epiCausality/epiCode/differential_methyl/modkit_output/single_base_day28_CRoff_vs_Unedit_dmr_20251109.bed
Rows,cols: (111263, 11)


/tmp/ipykernel_105915/2984644142.py:32: FutureWarning: errors='ignore' is deprecated and will raise in a future version. Use to_numeric without passing `errors` and catch exceptions explicitly instead
  df[c] = pd.to_numeric(df[c], errors='ignore')


chr1  1839174  1839175  chr1:1839174-1839175    0  m:1  1  .  0.1  \
0  chr1  1839205  1839206  chr1:1839205-1839206  0.0  m:1  1  .    0   
1  chr1  2014307  2014308  chr1:2014307-2014308  0.0  m:1  1  .    0   
2  chr1  2014312  2014313  chr1:2014312-2014313  0.0  m:1  1  .    0   
3  chr1  2014495  2014496  chr1:2014495-2014496  0.0  m:1  1  .    0   
4  chr1  2014703  2014704  chr1:2014703-2014704  0.0  m:1  1  .    0   

   m:100.00 ..1  
0  m:100.00   .  
1  m:100.00   .  
2  m:100.00   .  
3  m:100.00   .  
4  m:100.00   .

count  unique                     top   freq  \
chr1                    111263      25                    chr1  10199   
1839174               111263.0     NaN                     NaN    NaN   
1839175               111263.0     NaN                     NaN    NaN   
chr1:1839174-1839175    111263  111263  chrY:60366108-60366109      1   
0                     111263.0     NaN                     NaN    NaN   
m:1                     111263      30                     m:1  84432   
1                     111263.0     NaN                     NaN    NaN   
.                       111263      17                       .  98843   
0.1                   111263.0     NaN                     NaN    NaN   
m:100.00                111263      97                m:100.00  84924   
..1                     111263      48                       .  98843   

                                 mean             std       min         25%  \
chr1                              NaN             NaN       NaN         NaN   
1839174               67038412.499924  58278252.06657    1965.0  17183386.5   
1839175               67038413.499924  58278252.06657    1966.0  17183387.5   
chr1:1839174-1839175              NaN             NaN       NaN         NaN   
0                            0.023171        0.396716 -0.340835         0.0   
m:1                               NaN             NaN       NaN         NaN   
1                            1.074589        0.971748       1.0         1.0   
.                                 NaN             NaN       NaN         NaN   
0.1                          0.147866        0.666812       0.0         0.0   
m:100.00                          NaN             NaN       NaN         NaN   
..1                               NaN             NaN       NaN         NaN   

                             50%          75%          max  
chr1                         NaN          NaN          NaN  
1839174               50978348.0  102522914.5  247909190.0  
1839175               50978349.0  102522915.5  247909191.0  
chr1:1839174-1839175         NaN          NaN          NaN  
0                            0.0          0.0    20.662811  
m:1                          NaN          NaN          NaN  
1                            1.0          1.0        100.0  
.                            NaN          NaN          NaN  
0.1                          0.0          0.0         81.0  
m:100.00                     NaN          NaN          NaN  
..1                          NaN          NaN          NaN

Saved /home/michalula/code/epiCausality/epiCode/differential_methyl/modkit_output/dmr_hist_1839174.html
Saved /home/michalula/code/epiCausality/epiCode/differential_methyl/modkit_output/dmr_hist_1839175.html
Saved /home/michalula/code/epiCausality/epiCode/differential_methyl/modkit_output/dmr_hist_0.html
Saved /home/michalula/code/epiCausality/epiCode/differential_methyl/modkit_output/dmr_hist_1.html
Saved /home/michalula/code/epiCausality/epiCode/differential_methyl/modkit_output/dmr_hist_0.1.html


chr1    1839174    1839175      chr1:1839174-1839175    0  m:1  1  .  \
10198  chr1  247909190  247909191  chr1:247909190-247909191  0.0  m:1  1  .   
10197  chr1  247908237  247908238  chr1:247908237-247908238  0.0  m:1  1  .   
10196  chr1  247908067  247908068  chr1:247908067-247908068  0.0  m:0  1  .   
10195  chr1  247908029  247908030  chr1:247908029-247908030  0.0  m:1  1  .   
10194  chr1  247907834  247907835  chr1:247907834-247907835  0.0  m:1  1  .   
10193  chr1  247907773  247907774  chr1:247907773-247907774  0.0  m:1  1  .   
10192  chr1  247907753  247907754  chr1:247907753-247907754  0.0  m:0  1  .   
10191  chr1  247906959  247906960  chr1:247906959-247906960  0.0  m:1  1  .   
10190  chr1  247905815  247905816  chr1:247905815-247905816  0.0  m:1  1  .   
10189  chr1  247905805  247905806  chr1:247905805-247905806  0.0  m:1  1  .   
10188  chr1  247905731  247905732  chr1:247905731-247905732  0.0  m:1  1  .   
10187  chr1  247905460  247905461  chr1:247905460-247905461  0.0  m:1  1  .   
10186  chr1  247905020  247905021  chr1:247905020-247905021  0.0  m:0  1  .   
10185  chr1  247904442  247904443  chr1:247904442-247904443  0.0  m:1  1  .   
10184  chr1  247904196  247904197  chr1:247904196-247904197  0.0  m:1  1  .   
10183  chr1  247904170  247904171  chr1:247904170-247904171  0.0  m:0  1  .   
10182  chr1  247904138  247904139  chr1:247904138-247904139  0.0  m:0  1  .   
10181  chr1  247904110  247904111  chr1:247904110-247904111  0.0  m:1  1  .   
10180  chr1  247904049  247904050  chr1:247904049-247904050  0.0  m:1  1  .   
10179  chr1  247903999  247904000  chr1:247903999-247904000  0.0  m:1  1  .   
10178  chr1  247903787  247903788  chr1:247903787-247903788  0.0  m:1  1  .   
10177  chr1  247903711  247903712  chr1:247903711-247903712  0.0  m:1  1  .   
10176  chr1  247903664  247903665  chr1:247903664-247903665  0.0  m:1  1  .   
10175  chr1  247903650  247903651  chr1:247903650-247903651  0.0  m:0  1  .   
10174  chr1  247903560  247903561  chr1:247903560-247903561  0.0  m:1  1  .   
10173  chr1  247903543  247903544  chr1:247903543-247903544  0.0  m:1  1  .   
10172  chr1  247902681  247902682  chr1:247902681-247902682  0.0  m:1  1  .   
10171  chr1  247902678  247902679  chr1:247902678-247902679  0.0  m:1  1  .   
10170  chr1  247902278  247902279  chr1:247902278-247902279  0.0  m:0  1  .   
10169  chr1  247901928  247901929  chr1:247901928-247901929  0.0  m:1  1  .   
10168  chr1  247901430  247901431  chr1:247901430-247901431  0.0  m:1  1  .   
10167  chr1  247901377  247901378  chr1:247901377-247901378  0.0  m:1  1  .   
10166  chr1  247901336  247901337  chr1:247901336-247901337  0.0  m:1  1  .   
10165  chr1  247901281  247901282  chr1:247901281-247901282  0.0  m:1  1  .   
10164  chr1  247901253  247901254  chr1:247901253-247901254  0.0  m:1  1  .   
10163  chr1  247901161  247901162  chr1:247901161-247901162  0.0  m:0  1  .   
10162  chr1  247900627  247900628  chr1:247900627-247900628  0.0  m:1  1  .   
10161  chr1  247900509  247900510  chr1:247900509-247900510  0.0  m:1  1  .   
10160  chr1  247336961  247336962  chr1:247336961-247336962  0.0  m:1  1  .   
10159  chr1  247336959  247336960  chr1:247336959-247336960  0.0  m:1  1  .   
10158  chr1  247336884  247336885  chr1:247336884-247336885  0.0  m:1  1  .   
10157  chr1  247336842  247336843  chr1:247336842-247336843  0.0  m:1  1  .   
10156  chr1  247336838  247336839  chr1:247336838-247336839  0.0  m:1  1  .   
10155  chr1  247336613  247336614  chr1:247336613-247336614  0.0  m:1  1  .   
10154  chr1  247336386  247336387  chr1:247336386-247336387  0.0  m:1  1  .   
10153  chr1  247336362  247336363  chr1:247336362-247336363  0.0  m:1  1  .   
10152  chr1  247336146  247336147  chr1:247336146-247336147  0.0  m:0  1  .   
10151  chr1  247336138  247336139  chr1:247336138-247336139  0.0  m:1  1  .   
10150  chr1  247336068  247336069  chr1:247336068-247336069  0.0  m:1  1  .   
10149  chr1  247335904  247335905  chr1:247335904-2473359

Done. Visualizations saved to: /home/michalula/code/epiCausality/epiCode/differential_methyl/modkit_output
